src/index.js

```diff
@@ -60,33 +60,98 @@ const getVersion = memoize(async () => (await getPackage()).version || 'n/a');
 
 const getName = memoize(async () => (await getPackage()).name || 'n/a');
 
+async function uricheck(key, uri, timeout) {
+  const version = await getVersion();
+  const name = await getName();
 
-async function report(checks = {}, timeout = 10000, decorator = { body: xml, mime: 'application/xml', name: 'pingdom_http_custom_check' }) {
-  const start = Date.now();
+  const response = await request({
+    uri,
+    resolveWithFullResponse: true,
+    time: true,
+    timeout,
+    headers: {
+      'user-agent': `helix-status/${pkgversion} (${name}@${version})`,
+    },
+  });
+  return {
+    key,
+    response,
+  };
+}
+
+function seal(obj = {}, init, params) {
+  return Object.keys(obj).reduce((p, key) => {
+    const value = obj[key];
+    if (typeof value === 'function') {
+      // eslint-disable-next-line no-param-reassign
+      p[key] = value(params);
+    } else {
+      // eslint-disable-next-line no-param-reassign
+      p[key] = value;
+    }
+    return p;
+  }, init);
+}
+
+async function requestcheck(key, opts, timeout, params) {
   const version = await getVersion();
   const name = await getName();
 
-  try {
-    const checker = async ([key, uri]) => {
-      const response = await request({
-        uri,
-        resolveWithFullResponse: true,
-        time: true,
-        timeout,
-        headers: {
-          'user-agent': `helix-status/${pkgversion} (${name}@${version})`,
-        },
-      });
-      return {
-        key,
-        response,
-      };
-    };
+  const requestoptions = seal(opts, {
+    resolveWithFullResponse: true,
+    time: true,
+    timeout,
+  }, params);
+
+  const headers = seal(opts.headers, {
+    'user-agent': `helix-status/${pkgversion} (${name}@${version})`,
+  }, params);
+  requestoptions.headers = headers;
+
+  const response = await request(requestoptions);
+  return {
+    key,
+    response,
+  };
+}
 
+async function funccheck(key, func, params) {
+  const start = Date.now();
+  const result = await func(params);
+  const end = Date.now();
+
+  return {
+    key,
+    response: {
+      result,
+      timings: {
+        end: end - start,
+      },
+    },
+  };
+}
+
+function makechecker(timeout, params) {
+  return async function checker([key, check]) {
+    if (typeof check === 'string') {
+      return uricheck(key, check, timeout);
+    } else if (typeof check === 'function') {
+      return funccheck(key, check, params);
+    }
+    return requestcheck(key, check, timeout, params);
+  };
+}
+
+
+async function report(checks = {}, params, timeout = 10000, decorator = { body: xml, mime: 'application/xml', name: 'pingdom_http_custom_check' }) {
```

Comment:
Should we make `params` the last argument to keep compatibility with the original signature?